File: android-core/src/main/java/com/mparticle/internal/MessageBatch.java
package com.mparticle.internal;

import static com.mparticle.internal.Constants.MessageKey.DATA_PLAN_CONTEXT;
import static com.mparticle.internal.Constants.MessageKey.DATA_PLAN_ID;
import static com.mparticle.internal.Constants.MessageKey.DATA_PLAN_KEY;
import static com.mparticle.internal.Constants.MessageKey.DATA_PLAN_VERSION;

import com.mparticle.BuildConfig;
import com.mparticle.MParticle;
import com.mparticle.consent.CCPAConsent;
import com.mparticle.consent.ConsentInstance;
import com.mparticle.consent.ConsentState;
import com.mparticle.consent.GDPRConsent;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Map;
import java.util.UUID;

public class MessageBatch extends JSONObject {
    private long messageLengthBytes;

    protected MessageBatch() {
        super();
    }

    public static MessageBatch create(boolean history, ConfigManager configManager, JSONObject cookies, BatchId batchId) throws JSONException {
        MessageBatch uploadMessage = new MessageBatch();
        if (BuildConfig.MP_DEBUG) {
            uploadMessage.put(Constants.MessageKey.ECHO, true);
        }
        uploadMessage.put(Constants.MessageKey.TYPE, Constants.MessageType.REQUEST_HEADER);
        uploadMessage.put(Constants.MessageKey.ID, UUID.randomUUID().toString());
        uploadMessage.put(Constants.MessageKey.TIMESTAMP, System.currentTimeMillis());
        uploadMessage.put(Constants.MessageKey.MPARTICLE_VERSION, Constants.MPARTICLE_VERSION);
        uploadMessage.put(Constants.MessageKey.OPT_OUT_HEADER, configManager.getOptedOut());
        uploadMessage.put(Constants.MessageKey.CONFIG_UPLOAD_INTERVAL, configManager.getUploadInterval() / 1000);
        uploadMessage.put(Constants.MessageKey.MPARTICLE_CONFIG_VERSION, configManager.getEtag());
        uploadMessage.put(Constants.MessageKey.CONFIG_SESSION_TIMEOUT, configManager.getSessionTimeout() / 1000);
        uploadMessage.put(Constants.MessageKey.MPID, String.valueOf(batchId.getMpid()));
        uploadMessage.put(Constants.MessageKey.SANDBOX, configManager.getEnvironment().equals(MParticle.Environment.Development));
        uploadMessage.put(Constants.MessageKey.DEVICE_APPLICATION_STAMP, configManager.getDeviceApplicationStamp());

        if (history) {
            String deletedAttr = configManager.getUserStorage(batchId.getMpid()).getDeletedUserAttributes();
            if (deletedAttr != null) {
                uploadMessage.put(Constants.MessageKey.DELETED_USER_ATTRIBUTES, new JSONArray(deletedAttr));
                configManager.getUserStorage().deleteDeletedUserAttributes();
            }
        }

        uploadMessage.put(Constants.MessageKey.COOKIES, cookies);
        uploadMessage.put(Constants.MessageKey.PROVIDER_PERSISTENCE, configManager.getProviderPersistence());
        uploadMessage.put(Constants.MessageKey.INTEGRATION_ATTRIBUTES, configManager.getIntegrationAttributes());
        uploadMessage.addConsentState(configManager.getConsentState(batchId.getMpid()));
        uploadMessage.addDataplanContext(batchId.getDataplanId(), batchId.getDataplanVersion());
        return uploadMessage;
    }

    public void addConsentState(ConsentState consentState) {
        if (consentState != null) {
            try {
                JSONObject state = new JSONObject();
                this.put(Constants.MessageKey.CONSENT_STATE, state);


                Map<String, GDPRConsent> gdprState = consentState.getGDPRConsentState();
                if (gdprState != null) {
                    JSONObject gdpr = new JSONObject();
                    state.put(Constants.MessageKey.CONSENT_STATE_GDPR, gdpr);
                    for (Map.Entry<String, GDPRConsent> entry : gdprState.entrySet()) {
                        ConsentInstance consent = entry.getValue();
                        if (consent != null) {
                            addConsentStateJSON(gdpr, entry.getKey(), entry.getValue());
                        }
                    }
                }
                CCPAConsent ccpaConsent = consentState.getCCPAConsentState();
                if (ccpaConsent != null) {
                    JSONObject ccpa = new JSONObject();
                    state.put(Constants.MessageKey.CONSENT_STATE_CCPA, ccpa);
                    addConsentStateJSON(ccpa, Constants.MessageKey.CCPA_CONSENT_KEY, ccpaConsent);
                }
            } catch (JSONException ignored) {
            }
        }
    }

    public void addDataplanContext(String dataplanId, Integer dataplanVersion) throws JSONException {
        if (dataplanId != null) {
            JSONObject dataplan = new JSONObject();
            dataplan.put(DATA_PLAN_ID, dataplanId);
            if (dataplanVersion != null) {
                dataplan.put(DATA_PLAN_VERSION, dataplanVersion);
            }

            put(DATA_PLAN_CONTEXT, new JSONObject().put(DATA_PLAN_KEY, dataplan));
        }
    }

    public void addMessage(JSONObject message) {
        try {
            if (!has(Constants.MessageKey.MESSAGES)) {
                put(Constants.MessageKey.MESSAGES, new JSONArray());
            }
            getJSONArray(Constants.MessageKey.MESSAGES).put(message);
        } catch (JSONException ignored) {
        }
    }

    public void addReportingMessage(JSONObject reportingMessage) {
        try {
            if (!has(Constants.MessageKey.REPORTING)) {
                put(Constants.MessageKey.REPORTING, new JSONArray());
            }
            getJSONArray(Constants.MessageKey.REPORTING).put(reportingMessage);
        } catch (JSONException ignored) {
        }
    }

    public void setAppInfo(JSONObject appInfo) {
        try {
            put(Constants.MessageKey.APP_INFO, appInfo);
        } catch (JSONException ignored) {
        }
    }

    public void setDeviceInfo(JSONObject deviceInfo) {
        try {
            put(Constants.MessageKey.DEVICE_INFO, deviceInfo);
        } catch (JSONException ignored) {
        }
    }

    public JSONObject getAppInfo() {
        try {
            return getJSONObject(Constants.MessageKey.APP_INFO);
        } catch (JSONException e) {
            return null;
        }
    }

    public JSONObject getDeviceInfo() {
        try {
            return getJSONObject(Constants.MessageKey.DEVICE_INFO);
        } catch (JSONException e) {
            return null;
        }
    }

    public JSONArray getMessages() {
        try {
            return getJSONArray(Constants.MessageKey.MESSAGES);
        } catch (JSONException e) {
            return null;
        }
    }

    public void setIdentities(JSONArray identities) {
        try {
            put(Constants.MessageKey.USER_IDENTITIES, identities);
        } catch (JSONException ignored) {
        }
    }

    public void setUserAttributes(JSONObject userAttributes) {
        try {
            put(Constants.MessageKey.USER_ATTRIBUTES, userAttributes);
        } catch (JSONException ignored) {
        }
    }

    public long getMessageLengthBytes() {
        return messageLengthBytes;
    }

    public void incrementMessageLengthBytes(long bytes) {
        messageLengthBytes = messageLengthBytes + bytes;
    }

    private void addConsentStateJSON(JSONObject parentJSON, String key, ConsentInstance consentInstance) throws JSONException {
        JSONObject consentInstanceJSON = new JSONObject();
        parentJSON.put(key, consentInstanceJSON);
        consentInstanceJSON.put(Constants.MessageKey.CONSENT_STATE_CONSENTED, consentInstance.isConsented());
        if (consentInstance.getDocument() != null) {
            consentInstanceJSON.put(Constants.MessageKey.CONSENT_STATE_DOCUMENT, consentInstance.getDocument());
        }
        consentInstanceJSON.put(Constants.MessageKey.CONSENT_STATE_TIMESTAMP, consentInstance.getTimestamp());
        if (consentInstance.getLocation() != null) {
            consentInstanceJSON.put(Constants.MessageKey.CONSENT_STATE_LOCATION, consentInstance.getLocation());
        }
        if (consentInstance.getHardwareId() != null) {
            consentInstanceJSON.put(Constants.MessageKey.CONSENT_STATE_HARDWARE_ID, consentInstance.getHardwareId());
        }
    }
}

File: android-core/src/main/java/com/mparticle/internal/MessageBatch.kt
package com.mparticle.internal

import com.mparticle.BuildConfig
import com.mparticle.MParticle
import com.mparticle.consent.ConsentInstance
import com.mparticle.consent.ConsentState
import com.mparticle.internal.Constants.MessageKey
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.UUID

open class MessageBatch protected constructor() : JSONObject() {
    var messageLengthBytes: Long = 0
        private set

    fun addConsentState(consentState: ConsentState?) {
        consentState?.let {
            try {
                val state = JSONObject()
                this.put(MessageKey.CONSENT_STATE, state)

                val gdprState = consentState.gdprConsentState
                gdprState?.let {
                    val gdpr = JSONObject()
                    state.put(MessageKey.CONSENT_STATE_GDPR, gdpr)
                    for ((key, consent) in gdprState) {
                        consent?.let {
                            addConsentStateJSON(gdpr, key, consent)
                        }
                    }
                }
                val ccpaConsent = consentState.ccpaConsentState
                ccpaConsent ?.let {
                    val ccpa = JSONObject()
                    state.put(MessageKey.CONSENT_STATE_CCPA, ccpa)
                    addConsentStateJSON(ccpa, MessageKey.CCPA_CONSENT_KEY, ccpaConsent)
                }
            } catch (ignored: JSONException) {
            }
        }
    }

    @Throws(JSONException::class)
    fun addDataplanContext(dataplanId: String?, dataplanVersion: Int?) {
        dataplanId?.let {
            val dataplan = JSONObject()
            dataplan.put(MessageKey.DATA_PLAN_ID, dataplanId)
            dataplanVersion?.let {
                dataplan.put(MessageKey.DATA_PLAN_VERSION, dataplanVersion)
            }

            put(MessageKey.DATA_PLAN_CONTEXT, JSONObject().put(MessageKey.DATA_PLAN_KEY, dataplan))
        }
    }

    fun addMessage(message: JSONObject?) {
        try {
            if (!has(MessageKey.MESSAGES)) {
                put(MessageKey.MESSAGES, JSONArray())
            }
            getJSONArray(MessageKey.MESSAGES).put(message)
        } catch (ignored: JSONException) {
        }
    }

    fun addReportingMessage(reportingMessage: JSONObject?) {
        try {
            if (!has(MessageKey.REPORTING)) {
                put(MessageKey.REPORTING, JSONArray())
            }
            getJSONArray(MessageKey.REPORTING).put(reportingMessage)
        } catch (ignored: JSONException) {
        }
    }

    var appInfo: JSONObject?
        get() = try {
            getJSONObject(MessageKey.APP_INFO)
        } catch (e: JSONException) {
            null
        }
        set(appInfo) {
            try {
                put(MessageKey.APP_INFO, appInfo)
            } catch (ignored: JSONException) {
            }
        }

    var deviceInfo: JSONObject?
        get() {
            return try {
                getJSONObject(MessageKey.DEVICE_INFO)
            } catch (e: JSONException) {
                null
            }
        }
        set(deviceInfo) {
            try {
                put(MessageKey.DEVICE_INFO, deviceInfo)
            } catch (ignored: JSONException) {
            }
        }

    val messages: JSONArray?
        get() {
            return try {
                getJSONArray(MessageKey.MESSAGES)
            } catch (e: JSONException) {
                null
            }
        }

    fun setIdentities(identities: JSONArray?) {
        try {
            put(MessageKey.USER_IDENTITIES, identities)
        } catch (ignored: JSONException) {
        }
    }

    fun setUserAttributes(userAttributes: JSONObject?) {
        try {
            put(MessageKey.USER_ATTRIBUTES, userAttributes)
        } catch (ignored: JSONException) {
        }
    }

    fun incrementMessageLengthBytes(bytes: Long) {
        messageLengthBytes = messageLengthBytes + bytes
    }

    @Throws(JSONException::class)
    private fun addConsentStateJSON(parentJSON: JSONObject, key: String, consentInstance: ConsentInstance) {
        val consentInstanceJSON = JSONObject()
        parentJSON.put(key, consentInstanceJSON)
        consentInstanceJSON.put(MessageKey.CONSENT_STATE_CONSENTED, consentInstance.isConsented)
        consentInstance.document?.let {
            consentInstanceJSON.put(MessageKey.CONSENT_STATE_DOCUMENT, consentInstance.document)
        }
        consentInstanceJSON.put(MessageKey.CONSENT_STATE_TIMESTAMP, consentInstance.timestamp)
        consentInstance.location ?.let {
            consentInstanceJSON.put(MessageKey.CONSENT_STATE_LOCATION, consentInstance.location)
        }
        consentInstance.hardwareId?.let {
            consentInstanceJSON.put(MessageKey.CONSENT_STATE_HARDWARE_ID, consentInstance.hardwareId)
        }
    }

    companion object {
        @JvmStatic
        @Throws(JSONException::class)
        fun create(history: Boolean, configManager: ConfigManager, cookies: JSONObject?, batchId: BatchId): MessageBatch {
            val uploadMessage = MessageBatch()
            if (BuildConfig.MP_DEBUG) {
                uploadMessage.put(MessageKey.ECHO, true)
            }
            uploadMessage.put(MessageKey.TYPE, Constants.MessageType.REQUEST_HEADER)
            uploadMessage.put(MessageKey.ID, UUID.randomUUID().toString())
            uploadMessage.put(MessageKey.TIMESTAMP, System.currentTimeMillis())
            uploadMessage.put(MessageKey.MPARTICLE_VERSION, Constants.MPARTICLE_VERSION)
            uploadMessage.put(MessageKey.OPT_OUT_HEADER, configManager.optedOut)
            uploadMessage.put(MessageKey.CONFIG_UPLOAD_INTERVAL, configManager.uploadInterval / 1000)
            uploadMessage.put(MessageKey.MPARTICLE_CONFIG_VERSION, configManager.etag)
            uploadMessage.put(MessageKey.CONFIG_SESSION_TIMEOUT, configManager.sessionTimeout / 1000)
            uploadMessage.put(MessageKey.MPID, batchId.mpid.toString())
            uploadMessage.put(MessageKey.SANDBOX, ConfigManager.getEnvironment() == MParticle.Environment.Development)
            uploadMessage.put(MessageKey.DEVICE_APPLICATION_STAMP, configManager.deviceApplicationStamp)

            if (history) {
                val deletedAttr = configManager.getUserStorage(batchId.mpid).deletedUserAttributes
                deletedAttr ?.let {
                    uploadMessage.put(MessageKey.DELETED_USER_ATTRIBUTES, JSONArray(deletedAttr))
                    configManager.userStorage.deleteDeletedUserAttributes()
                }
            }

            uploadMessage.put(MessageKey.COOKIES, cookies)
            uploadMessage.put(MessageKey.PROVIDER_PERSISTENCE, configManager.providerPersistence)
            uploadMessage.put(MessageKey.INTEGRATION_ATTRIBUTES, configManager.integrationAttributes)
            uploadMessage.addConsentState(configManager.getConsentState(batchId.mpid))
            uploadMessage.addDataplanContext(batchId.dataplanId, batchId.dataplanVersion)
            return uploadMessage
        }
    }
}
